### PR TITLE
[GEN]: Introduce GEN dialect in Triton - part 6

### PR DESCRIPTION
**Pull request by etiotto**
Add conversion for GEN::FpToFp.

### CODE REVIEW

File: test/GEN/gen.mlir [hunk 1]
```diff
@@ -67,3 +67,24 @@ llvm.func @gen.sub_group_shuffle() {
...
+  // CHECK-NEXT: %6 = gen.conv.fptofp %arg1 {roundingMode = rtp} : f16 to f32
+  // CHECK-NEXT: %7 = gen.conv.fptofp %arg1 {roundingMode = rtz} : f16 to f32
+  %0 = gen.conv.fptofp %a {roundingMode = rte} : f32 to f16
+  %1 = gen.conv.fptofp %a {roundingMode = rtn} : f32 to f16
+  %2 = gen.conv.fptofp %a {roundingMode = rtp} : f32 to f16
+  %3 = gen.conv.fptofp %a {roundingMode = rtz} : f32 to f16
+  %4 = gen.conv.fptofp %b {roundingMode = rte} : f16 to f32
+  %5 = gen.conv.fptofp %b {roundingMode = rtn} : f16 to f32
+  %6 = gen.conv.fptofp %b {roundingMode = rtp} : f16 to f32
+  %7 = gen.conv.fptofp %b {roundingMode = rtz} : f16 to f32
```
Comment (whitneywhtsang) [inline test/GEN/gen.mlir:88]: can we remove roundingMode for one of them for testing.

File: third_party/intel/lib/GENToLLVM/GENToLLVMPass.cpp [hunk 2]
```diff
@@ -382,7 +443,8 @@ void mlir::triton::populateGENToLLVMConversionPatterns(
                GENGridDimXLowering, GENGridDimYLowering, GENGridDimZLowering>(
       converter);
   // clang-format on
-  patterns.add<GENBarrierLowering, SubGroupShuffleLowering>(converter);
+  patterns.add<GENBarrierLowering, GENSubGroupShuffleLowering>(converter);
```
[the remark this answers is not in the document]
Comment (etiotto) [reply third_party/intel/lib/GENToLLVM/GENToLLVMPass.cpp]: Just personal preference

File: test/TritonGEN/tritongen-invalid.mlir [hunk 3]
```diff
@@ -0,0 +1,23 @@
+// RUN: triton-opt -split-input-file -verify-diagnostics %s
+
+llvm.func @triton_gen.fptofp(%a : i32) {
+  // expected-error @+1 {{custom op 'triton_gen.conv.fptofp' invalid kind of type specified}}
+  %0 = triton_gen.conv.fptofp %a {roundingMode = rte} : i32 to i16
```
Comment (victor-eds) [inline test/TritonGEN/tritongen-invalid.mlir:5]:
```suggestion
  %0 = triton_gen.fconvert rte %a : i32 to i16
```

I would drop the `conv` prefix, rename the instruction (optional, but I think this follows convention in other dialects like `spirv.FConvert`), and drop the braces as that is usually used for the attribute dictionary.

File: test/TritonGEN/tritongen-invalid.mlir [hunk 4]
```diff
@@ -0,0 +1,23 @@
+// RUN: triton-opt -split-input-file -verify-diagnostics %s
+
+llvm.func @triton_gen.fptofp(%a : i32) {
```
Comment (victor-eds) [inline test/TritonGEN/tritongen-invalid.mlir:3]: I don't think we need this test as this is testing the MLIR type system itself. Our operation's type is `AnyFloat -> AnyFloat`

File: test/TritonGEN/tritongen-invalid.mlir [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+llvm.func @triton_gen.fptofp(%a : f32) {
+  // expected-error @+1 {{'triton_gen.conv.fptofp' op expecting first argument and result size to be different}}
+  %0 = triton_gen.conv.fptofp %a {roundingMode = rte} : f32 to f32
+  llvm.return
+}
+
+// -----
+
+llvm.func @triton_gen.fptofp(%a : f32) {
+  // expected-error @+1 {{'triton_gen.conv.fptofp' op expecting rounding mode for truncation}}
```
Comment (victor-eds) [inline test/TritonGEN/tritongen-invalid.mlir:20]: This asymmetry, asking for a rounding mode depending on the input types is a bit weird to me... We can discuss, but is choosing a default (`rte`?) a valid option?

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 6]
```diff
@@ -127,6 +127,51 @@ static LLVM::CallOp createSubGroupShuffle(ConversionPatternRewriter &rewriter,
...
+      break;
+    case TritonGEN::RoundingMode::RTP:
+      rounding = static_cast<int32_t>(llvm::RoundingMode::TowardPositive);
+      break;
+    case TritonGEN::RoundingMode::RTZ:
+      rounding = static_cast<int32_t>(llvm::RoundingMode::TowardZero);
+      break;
+    default:
+      llvm_unreachable("Unhandled rounding mode");
+    }
```
Comment (victor-eds) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:152]: MLIR offers a mechanism to convert attributes to different dialects. Please, add a TODO here so we remember to replace this switch with that.

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 7]
```diff
@@ -127,6 +127,51 @@ static LLVM::CallOp createSubGroupShuffle(ConversionPatternRewriter &rewriter,
...
+
+  Type argType = op.getArg().getType();
+  Type resType = op.getResult().getType();
+  unsigned resTySizeInBits = resType.getIntOrFloatBitWidth();
+  unsigned srcTySizeInBits = argType.getIntOrFloatBitWidth();
+  auto stringAttr =
+      (srcTySizeInBits > resTySizeInBits)
+          ? rewriter.getStringAttr("llvm.experimental.constrained.fptrunc")
+          : rewriter.getStringAttr("llvm.experimental.constrained.fpext");
```
Comment (victor-eds) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:161]:
```suggestion
          ? rewriter.getStringAttr(llvm::Intrinsic::getBaseName(llvm::Intrinsic::experimental_constrained_fptrunc))
          : rewriter.getStringAttr(llvm::Intrinsic::getBaseName(llvm::Intrinsic::experimental_constrained_fpext))
```

Comment (victor-eds) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:161]: Maybe add a TODO to add these to the `llvm` dialect as first class operations

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 8]
```diff
@@ -127,6 +127,51 @@ static LLVM::CallOp createSubGroupShuffle(ConversionPatternRewriter &rewriter,
...
+
+  if (rounding.has_value()) {
+    auto namedAttr = rewriter.getNamedAttr(
+        "roundingMode",
+        IntegerAttr::get(IntegerType::get(context, 32), rounding.value()));
+  }
+
+  // TODO: currently the LLVM dialect is unable to translate an intrinsic call
+  // with metadata correctly.
```
Comment (victor-eds) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:170]: Adding the intrinsics as operations to the dialect is the way to go IMO.

**Review by victor-eds — COMMENTED** [review]
Added comments to changes.

On a more conceptual topic. I think this operation should not be in our dialect. IMO we're exposing `arith`'s dialect lack of rounding mode capabilities. Adding them to the `arith` dialect might be the way to go in the future. WDYT?

File: test/GEN/gen.mlir (hunk 1 above)
Comment (etiotto) [reply test/GEN/gen.mlir]: done

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 7 above)
Comment (etiotto) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:161]: done

File: test/TritonGEN/tritongen-invalid.mlir (hunk 3 above)
Comment (etiotto) [reply test/TritonGEN/tritongen-invalid.mlir]: I dropped the `conv` prefix byt leaved the fptofp name intact.

File: test/TritonGEN/tritongen-invalid.mlir (hunk 4 above)
Comment (etiotto) [reply test/TritonGEN/tritongen-invalid.mlir:3]: Let's leave this test and remove it then in the GEN dialect for upstreaming.

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**